### PR TITLE
Add initial implementation of sprite-snapshot-tree

### PR DESCRIPTION
**Pull request by nickschot**
This is a data structure which:
- can classify and consolidate sprites into inserted, removed and kept sprite types
- contains classified sprites per _stable_ context (meaning this structure only contains _stable_ contexts and sprites are not necessarily handled by their closest ancestor context which would be in the process of inital render or final removal)

It currently does not handle:
- context aware bounds (snapshots of contexts and sprite-modifiers are taken).
- intermediateSprites (for interruptions)

We will need to add a step after this (probably not a concern of the sprite-snapshot-tree) which sets the correct ContextAwareBounds for all sprites.

This structure will feed into a refactored TransitionRunner in the animations service's maybeTransition(Task). We will be able to easily construct a scoped changeset per animating stable context as needed. The TransitionRunner and Changeset classes can likely be significantly simplified after we integrate this.

### CODE REVIEW

File: addon/addon/models/sprite-snapshot-node-builder.ts [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+        );
+
+        assert(
+          'kept sprite counterpart should have lastBounds and currentBounds',
+          counterpartModifier.lastBounds && counterpartModifier.currentBounds
+        );
+
+        // TODO: double check if we're setting the correct bounds here
+        if (counterpartModifier) {
+          sprite.counterpart.initialBounds = new ContextAwareBounds({
```
Comment (Aierie) [inline addon/addon/models/sprite-snapshot-node-builder.ts]: If it's an inserted sprite converted into a kept sprite, we need to set the sprite's initial bounds to the counterpart's initial bounds too.

File: addon/addon/models/sprite-snapshot-node-builder.ts [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+        SpriteType.Kept
+      );
+      keptSprite.counterpart = new Sprite(
+        removedSpriteModifier.element as HTMLElement,
+        removedSpriteModifier.id,
+        removedSpriteModifier.role,
+        SpriteType.Removed
+      );
+      spriteModifiers.add(insertedSpriteModifier);
+      spriteModifierToSpriteMap.set(insertedSpriteModifier, keptSprite);
```
Comment (Aierie) [inline addon/addon/models/sprite-snapshot-node-builder.ts:367]: After changing the inserted sprite into a kept sprite between contexts, overflow can crop out the animated counterpart that is animated as an orphan, if the inner context is managing the new kept sprite. 

https://user-images.githubusercontent.com/39579264/174632115-7984c4e3-effc-4c7a-af5a-67cb5567b31b.mov

Comment (nickschot) [reply addon/addon/models/sprite-snapshot-node-builder.ts:367]: This is interesting, we talked about taking the first shared ancestor context for kept sprites I think? Must not be working yet if this is happening 😅

File: addon/addon/models/sprite-snapshot-node-builder.ts (hunk 1 above)
Comment (nickschot) [reply addon/addon/models/sprite-snapshot-node-builder.ts]: I've got a bunch of modifications for this part locally too as I've now mostly integrated an updated version of the intermediate sprite system. We now also only create those for sprites that are actually animating which was not the case before.

Comment (nickschot) [reply addon/addon/models/sprite-snapshot-node-builder.ts]: We resolved this in the last commit.

**Pull request merged**